File: backend/services/html_parser.py
from bs4 import BeautifulSoup
from typing import List, Dict, Any


def parse_html_sections(html: str) -> List[Dict[str, Any]]:
    """
    Extrae secciones estructurales del HTML usando BeautifulSoup.
    Retorna una lista con metadata básica de cada sección detectada.
    """
    soup = BeautifulSoup(html, "html.parser")
    sections = []

    # Busca elementos candidatos a sección
    candidates = soup.find_all(
        ["section", "div", "header", "footer", "main", "article", "nav"],
        recursive=False,
    )

    # Si el body tiene hijos directos, usarlos
    body = soup.find("body")
    if body:
        candidates = list(body.children)
        candidates = [c for c in candidates if hasattr(c, "name") and c.name]

    # Filtrar solo elementos con suficiente contenido
    meaningful = []
    for el in candidates:
        text = el.get_text(strip=True)
        if len(text) > 20 or el.find(["img", "video", "picture"]):
            meaningful.append(el)

    for i, el in enumerate(meaningful):
        # Detectar imágenes
        imgs = [
            img.get("src", "") or img.get("data-src", "")
            for img in el.find_all("img")
            if img.get("src") or img.get("data-src")
        ]

        # Detectar clases de fondo
        classes = " ".join(el.get("class", []))
        style = el.get("style", "")
        has_bg_image = "background" in style.lower() and "url(" in style.lower()

        # Extraer color de fondo del style inline
        bg_color = "#ffffff"
        if "background-color" in style:
            import re
            match = re.search(r"background-color:\s*([^;]+)", style)
            if match:
                bg_color = match.group(1).strip()

        # Estimar padding
        padding_v = 64
        padding_h = 32
        if "py-" in classes or "pt-" in classes or "pb-" in classes:
            padding_v = 96
        if "px-" in classes or "pl-" in classes or "pr-" in classes:
            padding_h = 32

        # Determinar layout probable
        layout = "single"
        grid_classes = [c for c in el.get("class", []) if "grid" in c.lower()]
        flex_classes = [c for c in el.get("class", []) if "flex" in c.lower()]
        col_divs = el.find_all(attrs={"class": lambda x: x and any(
            "col" in c or "column" in c for c in x
        )}) if el.get("class") else []

        if grid_classes:
            layout = "grid-2"
        elif len(col_divs) >= 2:
            layout = "two-col"
        elif flex_classes:
            layout = "single"

        tag_name = el.name or "div"
        section_id = f"{tag_name}_{i}"
        if el.get("id"):
            section_id = el.get("id")
        elif el.get("class"):
            section_id = el.get("class")[0].replace("-", "_").lower()

        sections.append({
            "index": i,
            "tag": tag_name,
            "id_hint": section_id,
            "text_preview": el.get_text(strip=True)[:200],
            "images": [img for img in imgs if img],
            "has_background_image": has_bg_image,
            "background_color": bg_color,
            "layout_hint": layout,
            "suggested_padding_v": padding_v,
            "suggested_padding_h": padding_h,
            "html_snippet": str(el)[:500],
        })

    return sections

File: frontend/src/lib/server/htmlParser.ts
import { load } from 'cheerio'

export interface ParsedSection {
  index: number
  tag: string
  idHint: string
  textPreview: string
  images: string[]
  hasBackgroundImage: boolean
  backgroundColor: string
  layoutHint: string
  suggestedPaddingV: number
  suggestedPaddingH: number
  htmlSnippet: string
}

export function parseHtmlSections(html: string): ParsedSection[] {
  const $ = load(html)
  const sections: ParsedSection[] = []

  const candidates: ReturnType<typeof $>[] = []
  $('body').children().each((_, el) => { candidates.push($(el)) })

  // Fallback: si body está vacío, usar los hijos del root
  if (candidates.length === 0) {
    $('html').children().each((_, el) => { candidates.push($(el)) })
  }

  const meaningful = candidates.filter(($el) => {
    const text = $el.text().trim()
    return text.length > 20 || $el.find('img, video, picture').length > 0
  })

  meaningful.forEach(($el, i) => {
    const tag = ($el.prop('tagName') as string | undefined)?.toLowerCase() ?? 'div'

    // Imágenes
    const images: string[] = []
    $el.find('img').each((_, img) => {
      const src = $(img).attr('src') || $(img).attr('data-src')
      if (src) images.push(src)
    })

    // Fondo desde style inline
    const style = $el.attr('style') || ''
    const hasBgImage = /background/i.test(style) && /url\(/i.test(style)

    let bgColor = '#ffffff'
    const bgMatch = style.match(/background-color:\s*([^;]+)/i)
    if (bgMatch) bgColor = bgMatch[1].trim()

    // Padding sugerido desde clases
    const classes = $el.attr('class') || ''
    const padV = /py-|pt-|pb-/.test(classes) ? 96 : 64
    const padH = 32

    // Layout hint
    let layout = 'single'
    if (/grid/.test(classes)) layout = 'grid-2'
    else if ($el.find('[class*="col"]').length >= 2) layout = 'two-col'

    // ID hint
    const elId = $el.attr('id') || ''
    const firstClass = classes.split(' ').find((c) => c.trim()) || ''
    const idHint =
      elId ||
      firstClass.replace(/[^a-z0-9]/gi, '_').toLowerCase() ||
      `${tag}_${i}`

    sections.push({
      index: i,
      tag,
      idHint,
      textPreview: $el.text().trim().slice(0, 200),
      images,
      hasBackgroundImage: hasBgImage,
      backgroundColor: bgColor,
      layoutHint: layout,
      suggestedPaddingV: padV,
      suggestedPaddingH: padH,
      htmlSnippet: $.html($el[0] as Parameters<typeof $.html>[0])?.slice(0, 600) ?? '',
    })
  })

  return sections
}
